pypose/module/LQR.py

```diff
@@ -0,0 +1,235 @@
+        new_x, new_u, objs, tau = self.DP_LQR_forward(x_init, Q, p, Ks, ks)
+        return new_x, new_u, objs, tau
+
+    
+    def DP_LQR_backward(self, Q, p):
+        r'''
+        The backward recursion of LQR.
+
+        Args:
+            Q (:obj:`Tensor`): The matrix of quadratic parameter.
+            p (:obj:`Tensor`): The constant vector of quadratic parameter.
+
+        Returns:
+            Tuple of Tensor: The status feedback controller at all steps.
+        '''
+
+        Ks = []
+        ks = []
+        Vtp1 = vtp1 = None
+        for t in range(self.T-1, -1, -1): 
+            pt = p[t]
+            if pt.ndim == 2:
+                pt = pt.unsqueeze(1)
+            elif pt.ndim == 1:
+                pt = pt.unsqueeze(0)
+
+            if self.c1.ndim ==3:
+                c1 = self.c1[t]
+                c1 = c1.unsqueeze(1)
+            elif self.c1.ndim == 2:
+                c1 = self.c1.unsqueeze(1)
+            elif self.c1.ndim == 1:
+                c1 = self.c1.unsqueeze(0)
+
+            if self.system.A.ndim == 4:
+                A = self.system.A[t]
+                B = self.system.B[t]
+                F = torch.cat((A, B), axis = 2)
+            elif self.system.A.ndim == 3:
+                F = torch.cat((self.system.A, self.system.B), axis = 2)
+            elif self.system.A.ndim == 2:
+                F = torch.cat((self.system.A, self.system.B), dim = 1)
+
+            if t == self.T-1:
+                Qt = Q[t]
+                qt = pt.mT
+            else:
+                Ft = F
+                Ft_T = Ft.mT
+                Qt = Q[t] + Ft_T.matmul(Vtp1).matmul(Ft)
+                if self.c1 is None or self.c1.numel() == 0:
+                    qt = pt.mT + Ft_T.matmul(vtp1) 
+                else:
+                    ft = c1.mT
+                    qt = pt.mT + Ft_T.matmul(Vtp1).matmul(ft)+ Ft_T.matmul(vtp1)
+            
+            n_state = self.n_state
+            Qt_xx = Qt[:, :n_state, :n_state]
+            Qt_xu = Qt[:, :n_state, n_state:]
+            Qt_ux = Qt[:, n_state:, :n_state]
+            Qt_uu = Qt[:, n_state:, n_state:]
+            qt_x = qt[:, :n_state, :]
+            qt_u = qt[:, n_state:, :]
+
+            if self.n_ctrl == 1:
+                Kt = -(1./Qt_uu)*Qt_ux
+                kt = -(1./Qt_uu)*qt_u
+            else:
+                Qt_uu_inv = [torch.linalg.pinv(Qt_uu[i]) for i in range(Qt_uu.shape[0])]
+                Qt_uu_inv = torch.stack(Qt_uu_inv)
+                Kt = -Qt_uu_inv.bmm(Qt_ux)
+                kt = -Qt_uu_inv.matmul(qt_u)
+
+            Kt_T = Kt.mT
+
+            Ks.append(Kt)
+            ks.append(kt)
+
+            Vtp1 = Qt_xx + Qt_xu.matmul(Kt) + Kt_T.matmul(Qt_ux) + Kt_T.matmul(Qt_uu).matmul(Kt)
+            vtp1 = qt_x + Qt_xu.matmul(kt) + Kt_T.matmul(qt_u) + Kt_T.matmul(Qt_uu).matmul(kt)
+
+        return Ks, ks
+
+    def DP_LQR_forward(self, x_init, Q, p, Ks, ks):
+        r'''
+        The forward recursion of LQR.
```

Comment:
The forward recursion of dynamic programming to solve LQR. u_t = Kt*x_t + kt.  (We can explicitly provide this equation since it's simple enough.)